frontend_tests/node_tests/image_upload_widget.js

```diff
@@ -0,0 +1,90 @@
+zrequire('image_upload_widget');
+zrequire('avatar');
+zrequire('ui_report');
+set_global('$', global.make_zjquery());
+
+const _page_params = {
+    is_admin: true,
+    zulip_plan_is_not_limited: true,
+    avatar_source: 'G',
+};
+
+let form_data;
+
+const _FormData = function () {
+    return form_data;
+};
+
+const noop = () => {};
+
+const _realm_logo = {
+    build_realm_logo_widget: noop,
+};
+
+const _channel = {};
+
+const _realm_icon = {};
+
+set_global('channel', _channel);
+set_global('realm_icon', _realm_icon);
+set_global('realm_logo', _realm_logo);
+set_global('page_params', _page_params);
+set_global('FormData', _FormData);
+
+
+run_test('image_upload_widget', () => {
+
+    function test_image_upload(upload_icon, widget) {
+        form_data = {
+            append: function (field, val) {
+                form_data[field] = val;
+            },
+        };
+
+        const file_input = [
+            {files: ['image1.png', 'image2.png']},
+        ];
+
+        let posted;
+        let url;
+        let upload_success;
+        let upload_error;
+        if (widget === 'user-avatar-upload-widget') {
+            url = '/json/users/me/avatar';
+        } else if (widget === 'realm-icon-upload-widget') {
+            url = '/json/realm/icon';
+        } else {
+            url = '/json/realm/logo';
+        }
+        channel.post = function (req) {
+            posted = true;
+            assert.equal(req.url, url);
+            assert.equal(req.data.csrfmiddlewaretoken, 'fake-csrf-token');
+            assert.equal(req.data['file-0'], 'image1.png');
+            assert.equal(req.data['file-1'], 'image2.png');
+            upload_success = req.success;
+            upload_error = req.error;
+        };
+
+        upload_icon(file_input, widget);
+        upload_success();
+        upload_error();
```

Comment:
You are mocking server call here right ?
So line 65, 66, 70 and 71 is basically mocking both success and error at same time which is not possible. In reality channel.post will run either function passed in as `success` param or as `error` param. Also this is not right way to trigger these.

Reply:
```javascript
channel.post = function (req) {
            posted = true;
            assert.equal(req.url, url);
            assert.equal(req.data.csrfmiddlewaretoken, 'fake-csrf-token');
            assert.equal(req.data['file-0'], 'image1.png');
            assert.equal(req.data['file-1'], 'image2.png');
            // maybe add a delay here to simulate network request using setTimeout ? 
           //not sure I have never done this in zulip 
            req.success(); // simulate success server call
        };
```

```javascript
channel.post = function (req) {
            posted = true;
            assert.equal(req.url, url);
            assert.equal(req.data.csrfmiddlewaretoken, 'fake-csrf-token');
            assert.equal(req.data['file-0'], 'image1.png');
            assert.equal(req.data['file-1'], 'image2.png');
            // maybe add a delay here to simulate network request using setTimeout ? 
           //not sure I have never done this in zulip 
            req.error(); // simulate error server call, will trigger error codepath of function under test.
        };
```

Use both cases in separate test cases and separately. Don't call them together as you are doing in line 70 and 71

Reply:
I don't think we need 2 functions for `req.success` and `req.error` since the parameters are same for both the functions and the success and error functions will only change the UI elements eg `hide loading spinner`.